thread-manager/src/tokio_runtime.rs

```diff
@@ -0,0 +1,181 @@
+    pub worker_threads: usize,
+    ///max number of blocking threads tokio is allowed to spawn
+    pub max_blocking_threads: usize,
+    /// Priority in range 0..99
+    pub priority: u8,
+    pub policy: String,
+    pub stack_size_bytes: usize,
+    pub event_interval: u32,
+    pub core_allocation: CoreAllocation,
+}
+
+impl Default for TokioConfig {
+    fn default() -> Self {
+        Self {
+            core_allocation: CoreAllocation::OsDefault,
+            worker_threads: 8,
+            max_blocking_threads: 1,
+            priority: crate::policy::DEFAULT_PRIORITY,
+            policy: "OTHER".to_owned(),
+            stack_size_bytes: 2 * 1024 * 1024,
+            event_interval: 61,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TokioRuntime {
+    pub tokio: tokio::runtime::Runtime,
+    pub config: TokioConfig,
+    pub counters: Arc<ThreadCounters>,
+}
+
+impl Deref for TokioRuntime {
+    type Target = tokio::runtime::Runtime;
+
+    fn deref(&self) -> &Self::Target {
+        &self.tokio
+    }
+}
+
+impl TokioRuntime {
+    /// Starts the metrics sampling task on the runtime to monitor how many workers are busy doing useful things.
+    pub fn start_metrics_sampling(&self, period: Duration) {
+        let counters = self.counters.clone();
+        self.tokio.spawn(metrics_sampler(counters, period));
+    }
+
+    pub fn new(name: String, cfg: TokioConfig) -> anyhow::Result<Self> {
+        debug_assert!(name.len() < MAX_THREAD_NAME_CHARS, "Thread name too long");
+        let num_workers = if cfg.worker_threads == 0 {
+            num_cpus::get()
+        } else {
+            cfg.worker_threads
+        };
+        let chosen_cores_mask = cfg.core_allocation.as_core_mask_vector();
+
+        let base_name = name.clone();
+        let mut builder = match num_workers {
+            1 => tokio::runtime::Builder::new_current_thread(),
+            _ => {
+                let mut builder = tokio::runtime::Builder::new_multi_thread();
+                builder.worker_threads(num_workers);
+                builder
+            }
+        };
+        let atomic_id: AtomicUsize = AtomicUsize::new(0);
+
+        let counters = Arc::new(ThreadCounters {
+            namespace: format!("thread-manager-tokio-{}", &base_name).leak(), // no workaround, metrics crate will only consume 'static str
+            total_threads_cnt: cfg.worker_threads as u64,
+            active_threads_cnt: AtomicU64::new(
+                (num_workers.wrapping_add(cfg.max_blocking_threads)) as u64,
+            ),
+        });
+        let counters_clone1 = counters.clone();
+        let counters_clone2 = counters.clone();
+        builder
+            .event_interval(cfg.event_interval)
+            .thread_name_fn(move || {
+                let id = atomic_id.fetch_add(1, Ordering::Relaxed);
+                format!("{}-{}", base_name, id)
+            })
+            .on_thread_park(move || {
+                counters_clone1.on_park();
+            })
```

Comment:
nit/style - I'm quite favorable to doing local clones on these closure creations:
```suggestion
            .on_thread_park({
                let counters = counters.clone();
                move || {
                    counters.on_park();
                }
            })
```
limits scope and imo this makes the cloning more obvious than using a `X_clone` naming,

Reply:
Yes, this is indeed far nicer than my original implementation, thanks for the tip!